exec/java-exec/src/test/java/org/apache/drill/exec/physical/resultSet/impl/TestResultSetLoaderUnprojected.java

```diff
@@ -0,0 +1,455 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.drill.exec.physical.resultSet.impl;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+import java.util.List;
+
+import org.apache.drill.common.expression.SchemaPath;
+import org.apache.drill.common.types.TypeProtos.MinorType;
+import org.apache.drill.exec.physical.resultSet.ResultSetLoader;
+import org.apache.drill.exec.physical.resultSet.RowSetLoader;
+import org.apache.drill.exec.physical.resultSet.impl.ResultSetLoaderImpl.ResultSetOptions;
+import org.apache.drill.exec.physical.resultSet.project.Projections;
+import org.apache.drill.exec.physical.rowSet.RowSet;
+import org.apache.drill.exec.physical.rowSet.RowSet.SingleRowSet;
+import org.apache.drill.exec.physical.rowSet.RowSetTestUtils;
+import org.apache.drill.exec.record.metadata.SchemaBuilder;
+import org.apache.drill.exec.record.metadata.TupleMetadata;
+import org.apache.drill.exec.vector.accessor.ArrayWriter;
+import org.apache.drill.exec.vector.accessor.ScalarWriter;
+import org.apache.drill.exec.vector.accessor.TupleWriter;
+import org.apache.drill.test.SubOperatorTest;
+import org.apache.drill.test.rowSet.RowSetUtilities;
+import org.junit.Test;
+
+/**
+ * Verify the correct functioning of the "dummy" columns created
+ * for unprojected columns.
+ */
+public class TestResultSetLoaderUnprojected  extends SubOperatorTest {
```

Comment:
This test case provides a good use guide. In addition, will it be possible for us to remove Union completely from the data type in the future?

Reply:
The UNION type has been discussed for as long as I've been involved in the project: since 2016. The idea is simple: Drill should be able to read any kind of JSON, and UNION (plus LIST, etc.) have been essential for this. The problem, as we've also discussed for a long time, is that UNION barely works, and JDBC and similar clients can't make sense of it. That is, UNION turned out to be the wrong solution to the problem.

Still, there is always the hope that UNION can be made to work -- somehow. Also, there is always the worry that "maybe somebody, somewhere, is using it!

Removing UNION would require quite a bit of work: remove the vectors, EVF support, old-school vector writers, old-school JSON parser, operators that somewhat handle UNION, etc. What would be the advantage of doing so? Drill wouldn't be faster, though the code might be a bit smaller. There is no harm in leaving UNION in Drill. It doesn't cause any problems if it is not used. It only causes problems when it *is* used. Leaving it in Drill also avoids the long discussions that would otherwise occur if someone proposed to remove it.

Here's another idea. If folks really want the promise of (unstructured) JSON, then maybe open a Jira ticket asking to implement a Druid-like alternative: use Drill's OBJECT vector to store JSON as a set of Java objects (scalars, maps, and lists). Provide the needed serialization. Adopt the standard SQL JSON functions, since the SQL standards committee (and other vendors) have done the hard work to figure out a way to integrate JSON into SQL. Once that new, modern, JSON solution were available, we could easily remove the old UNION-based support with no worries.

Reply:
Thanks for this Paul! We must remember to squash when merging, we got the WIP commits from feature branch into master. We've all done something similar at some point-- I wonder if we should change our process to allow unprotecting the master branch temporarily (after agreement in the mailling list) to allow simple clean ups like squashing and dropping? From what I understand, Calcite works that way...

Reply:
> Thanks for this Paul! We must remember to squash when merging, we got the WIP commits from the feature branch into master. We've all done something similar at some point-- I wonder if we should change our process to allow unprotecting the master branch temporarily (after agreement in the mailling list) to allow simple clean ups like squashing and dropping? From what I understand, Calcite works that way...

At least there aren't my snarky commit messages...  ;-)

Reply:
P.S. I'm happy to live with the WIP commits in master too, just asking what folks think.

Reply:
My bad. My other project likes to leave these in master; I forgot Drill does not.

Since there is not much activity, I can squash the commits within the master branch and do a force push. Normally that is a big NO NO in active projects, but it should not actually cause problems here. I'll go ahead and do that tomorrow unless anyone objects.

Reply:
@paul-rogers, @cgivre  [commented that he's in favour of leaving master as is](https://github.com/apache/drill/pull/2866#issuecomment-1880409413) and I've since merged [a commit on top](https://github.com/apache/drill/commit/f5fb7f5a4023651252afb1f907311d71840eb144). I do think it would still be feasible for us to go back and squash (for exactly the reason you give) but at this point we could also just leave it where it is?

P.S. The process is a little laborious. A conventional commit switching off master branch protection in .asf.yaml, then the force push doing the clean up and switching master branch protection back on.

Reply:
Just to be clear... I was just saying that if this is a major headache and you don't want to deal with it, my vote is to leave it alone.  If it isn't a big headache and you want to, I have no issues there as well.